### PR TITLE
Added widget generator tool

### PR DESCRIPTION
**Pull request by Thidasala**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

**Comment by fazlan-nazeem** [comment]
we don't need the package-lock.json right?

File: product/distribution/src/assembly/bin.xml [hunk 1]
```diff
@@ -90,6 +90,11 @@
             <directory>carbon-home/resources/security</directory>
             <outputDirectory>resources/security</outputDirectory>
             <fileMode>644</fileMode>
+        </fileSet>
+          <fileSet>
+            <directory>carbon-home/generator-widgetgenerator</directory>
+            <outputDirectory>wso2/tools/generator-widgetgenerator</outputDirectory>
```
Comment (fazlan-nazeem) [inline product/distribution/src/assembly/bin.xml]: shall we change the name to **widget-generator**?

File: product/distribution/carbon-home/generator-widgetgenerator/app/templates/resources/locales/en.json [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{
+    "widget.heading": "<%= widgetHeading %>"
+}
+
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/app/templates/resources/locales/en.json]: new line required

File: product/distribution/carbon-home/generator-widgetgenerator/app/templates/resources/widgetConf-Rdbms.json [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+          "publishingLimit": 2147483647,
+          "isPurgingEnable": false
+        }
+      }
+    },
+    "options": [
+    ]
+  }
+}
+
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/app/templates/resources/widgetConf-Rdbms.json]: new line required

File: product/distribution/carbon-home/generator-widgetgenerator/README.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
+# Introduction
+
+## What is widget-generator tool?
+Widget generator tool generate customized widget tamplates based on user inputs. Users can modify the widget template and create new widgets which can be used across analytics dashboards.
+Tool generates three types of widget templates which supports both RDBMS Data Provider and SiddhiStore Data Provider.
+
+#### 1. Subscribesr Widgets
```
Comment (pahans) [inline product/distribution/carbon-home/generator-widgetgenerator/README.md]:
```suggestion
#### 1. Subscriber Widgets
```

**Comment by RAVEENSR** [comment]
Let's change the folder name as well? generator-widgetgenerator --> widget-generator

**Comment by fazlan-nazeem** [comment]
@RAVEENSR apparently there is a naming convention of yomen where the folder name has to start with 'generator'

**Comment by pahans** [comment]
> @RAVEENSR apparently there is a naming convention of yomen where the folder name has to start with 'generator'

IMO since we don't expose yeoman directly to widget developer and we do not intend to publish this to npm I think its okay to not to stick with the convention.

File: product/distribution/carbon-home/generator-widgetgenerator/README.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+#### 2. Publisher Widgets
+#### 3. Stand-alone Widgets
+
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/README.md]: Shall we mention here that the widget generated by default is a subscriber + Siddhi data provider widget

File: product/distribution/carbon-home/generator-widgetgenerator/README.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+>In order to use the sample Subscriber widget generated by the tool, User must use the DateTimeRangePicker widget as the publisher widget.
+
+
+# Quick Start Guide
+
+## Hoe to use the tool?
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/README.md]: typo 
s/Hoe/How

File: product/distribution/carbon-home/generator-widgetgenerator/README.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+1. Navigate to the <TOOL_HOME> directory and issue the following command to install YEOMAN globally.
+   - `$ sudo npm install -g yo`
+
+2. To link the global dependencies issue the following command.
+
+   - `$ sudo npm link`
+
+3. OPen a terminal in the <TOOL_HOME> directory and Use following commands to create your preffered widget skeleton.
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/README.md]: s/OPen/Open

File: product/distribution/carbon-home/generator-widgetgenerator/README.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+#### 2. Publisher Widgets
+#### 3. Stand-alone Widgets
+
+>In order to use the sample Subscriber widget generated by the tool, User must use the DateTimeRangePicker widget as the publisher widget.
+
+
+# Quick Start Guide
+
+## Hoe to use the tool?
+1. Navigate to the <TOOL_HOME> directory and issue the following command to install YEOMAN globally.
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/README.md]: <TOOL_HOME> is not widely known IMO. So shall we say relative to <ANALYTICS_HOME>?

File: product/distribution/carbon-home/generator-widgetgenerator/README.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+       `$ npm run createwidget`
+
+    - Issue the following command replacing the `argument1` and `argument2` as follows to create other widget types.
+    
+      `$npm run createwidget--widgettype:`argument1`--dataprovider:`argument2
+
+    - "argument1" is reffered to the widget type, replace the "argument1" with `subscriber`, `publisher` or `stand-alone` type.
+
+    - "argument2" is reffered to the data provider type, replace the "argument2" with `SiddhiDataProvider` or `RDBMSDataProvider` type.
+
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/README.md]: are these arguments case sensitive?

File: product/distribution/carbon-home/generator-widgetgenerator/README.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    - For example following command will create a publisher widget template with RDBMS Data Provider.
+
+      `$npm run createwidget--widgettype:publisher --dataprovider:RDBMSDataProvider`
+3. Enter your widget name and widget heading to the terminal.
+
+3. Based on the preferences, new widget skelotoon will be created inside the `<TOOL_HOME>/widgetTemplates` directory and the relevent dependencies will be installed to the `<TOOL_HOME>/widgetTemplates/<WIDGET_ROOT>` directory.
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/README.md]: s/skelotoon/skeleton

File: product/distribution/carbon-home/generator-widgetgenerator/README.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+4. Navigate to the `<WIDGET_ROOT>` directory and modify the file content to create your own widget.
+
+5. Navigate to the `<WIDGET_ROOT>` directory and Issue the following command to build the widget.
+      - `$npm run build`
+
+6. Restart the analytics dashboard profile to load the newly created widget to the dashboard. 
+
+7. Additionally, to start a node development environment and create a symlink from `<WIDGET_ROOT>` directory to `<HOME>/wso2/...widgets`  directory, Open a terminal in the `<WIDGET_ROOT>` directory and issue following commands.
+      - `$npm run dev`
+      - `$npm run create-symlink`
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/README.md]: Don't we need to have create-symlink [1] module installed to run the above command?
[1] https://www.npmjs.com/package/create-symlink

File: product/distribution/carbon-home/generator-widgetgenerator/app/index.js [hunk 12]
```diff
@@ -0,0 +1,201 @@
+var Generator = require('yeoman-generator');
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/app/index.js]: License missing?

File: product/distribution/carbon-home/generator-widgetgenerator/README.md (hunk 11 above)
Comment (pahans) [reply product/distribution/carbon-home/generator-widgetgenerator/README.md]: @TanyaM I think he is referring [here](https://github.com/wso2/analytics-apim/pull/1018/files#diff-3754187bf3c8517493898d27278b43feR188).

File: product/distribution/carbon-home/generator-widgetgenerator/app/index.js [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+        'webpack': '^4.41.2',
+        'rimraf': '^2.6.3',
+        'webpack-cli': '^3.3.10'
+      },
+      scripts: {
+        'build': 'node_modules/.bin/webpack -p',
+        "postbuild": "cp -r dist/APIMSample ../../../../dashboard/deployment/web-ui-apps/analytics-dashboard/extensions/widgets",
+        'clean': 'rimraf dist',
+        'dev': 'NODE_ENV=development node_modules/.bin/webpack -d --config webpack.config.js --watch --progress',
+        "symlink": "ln -s -r dist/APIMSample ../../../../dashboard/deployment/web-ui-apps/analytics-dashboard/extensions/widgets"
```
Comment (pahans) [inline product/distribution/carbon-home/generator-widgetgenerator/app/index.js]: `symlink` or `create-symlink`?. It says `create-symlink` in the doc.

File: product/distribution/carbon-home/generator-widgetgenerator/app/index.js [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+  writing() {
+    if (this.options.widgetType === 'subscriber') {
+      this.fs.copyTpl(
+        this.templatePath('subscriber.jsx'),
+        this.destinationPath(`widgetTemplates/${this.answers.widgetName}/src/${this.answers.widgetName}widget.jsx`),
+        { className: `${this.answers.widgetName}widget`,
+          widgetId: this.answers.widgetName,
+        }
+      );
+      if (this.options.dataProvider === 'SiddhiDataProvider') {
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/app/index.js]: shall we make this case insensitive?

File: product/distribution/carbon-home/generator-widgetgenerator/app/index.js [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+          this.templatePath('resources/widgetConf-Siddhi.json'),
+          this.destinationPath(`widgetTemplates/${this.answers.widgetName}/src/resources/widgetConf.json`),
+          { className: `${this.answers.widgetName}widget`,
+            widgetId: this.answers.widgetName,
+            widgetType: 'subscriber',
+          }
+        );
+      } else {
+        this.fs.copyTpl(
+          this.templatePath('resources/widgetConf-Rdbms.json'),
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/app/index.js]: Shall we use a convention for file names? Either camel case or seperated by hiphens in all simple case?

File: product/distribution/carbon-home/generator-widgetgenerator/app/index.js [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+          this.destinationPath(`widgetTemplates/${this.answers.widgetName}/src/resources/widgetConf.json`),
+          { className: `${this.answers.widgetName}widget`,
+            widgetId: this.answers.widgetName,
+            widgetType: 'subscriber',
+          }
+        );
+      }
+    }else if (this.options.widgetType == 'publisher'){
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/app/index.js]: formatting issue. There needs to be a space after } before else if
Have we run es-lint on this file? If not shall we run and correct the issues.

File: product/distribution/carbon-home/generator-widgetgenerator/app/index.js [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+      this.fs.copyTpl(
+        this.templatePath('publisher.jsx'),
+        this.destinationPath(`widgetTemplates/${this.answers.widgetName}/src/${this.answers.widgetName}widget.jsx`),
+        { className: `${this.answers.widgetName}widget`,
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/app/index.js]: formatting issue?

File: product/distribution/carbon-home/generator-widgetgenerator/README.md [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+#### 3. Stand-alone Widgets
+
+>In order to use the sample Subscriber widget generated by the tool, User must use the DateTimeRangePicker widget as the publisher widget.
+
+
+# Quick Start Guide
+
+## Hoe to use the tool?
+1. Navigate to the <TOOL_HOME> directory and issue the following command to install YEOMAN globally.
+   - `$ sudo npm install -g yo`
```
Comment (pahans) [inline product/distribution/carbon-home/generator-widgetgenerator/README.md]: (Discussed offline) to remove yoeman globally install step and move yoeman into dev dependencies.

File: product/distribution/carbon-home/generator-widgetgenerator/app/index.js [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+          widgetId: this.answers.widgetName
+        }
+      );
+      if (this.options.dataProvider == 'SiddhiDataProvider') {
+        this.fs.copyTpl(
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/app/index.js]: indentation issue. please fix else where

File: product/distribution/carbon-home/generator-widgetgenerator/app/index.js [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+          this.templatePath('resources/widgetConf-Rdbms.json'),
+          this.destinationPath(`widgetTemplates/${this.answers.widgetName}/src/resources/widgetConf.json`),
+          { className: `${this.answers.widgetName}widget`,
+            widgetId: this.answers.widgetName,
+            dataProvider: 'RDBMSStreamingDataProvider',
+            widgetType: 'publisher'
+          }
+        );
+      }
+    }else{
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/app/index.js]: formatting issue

File: product/distribution/carbon-home/generator-widgetgenerator/app/templates/publisher.jsx [hunk 21]
```diff
@@ -0,0 +1,333 @@
...
+                // TODO: Show error message.
+            });
+        });
+    }
+
+    componentDidMount() {
+        // Use this method to load default values to publisher parameters
+        super.publish('Initial Message');
+        const { widgetID } = this.props;
+        // This method is mandatory and used to load the providerConfig of widgetConf.json file.
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/app/templates/publisher.jsx]: Shall we rephrase the above like this?
This function retrieves the provider configuration defined in the widgetConf.json file and make it available to be used inside the widget.

File: product/distribution/carbon-home/generator-widgetgenerator/app/templates/publisher.jsx [hunk 22]
```diff
@@ -0,0 +1,333 @@
...
+        const { providerConfig } = this.state;
+        const { id, widgetID: widgetName } = this.props;
+
+        const dataProviderConfigs = cloneDeep(providerConfig);
+        dataProviderConfigs.configs.config.queryData.queryName = 'alertQuery';
+        dataProviderConfigs.configs.config.queryData.queryValues = {
+            '{{weekStart}}': (Date.now() - 604800),
+            '{{weekEnd}}': Date.now(),
+        };
+        super.getWidgetChannelManager()
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/app/templates/publisher.jsx]: Shall we add a comment here too?

File: product/distribution/carbon-home/generator-widgetgenerator/app/templates/publisher.jsx [hunk 23]
```diff
@@ -0,0 +1,333 @@
...
+    }
+
+    /**
+     * Formats data retrieved
+     * @param {object} message - data retrieved
+     * @memberof <%= className %>
+     * */
+    handleQueryResults(message) {
+        const { data } = message;
+        // Insert the code to handle the data recived through query
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/app/templates/publisher.jsx]: s/recived/received/g

File: product/distribution/carbon-home/generator-widgetgenerator/app/templates/publisher.jsx [hunk 24]
```diff
@@ -0,0 +1,333 @@
...
+    }
+
+    /**
+     * Publishing the parameters for the subscriber widgets
+     * @memberof <%= className %>
+     */
+    publishMsg() {
+        const { currentMessage } = this.state;
+        if (currentMessage && (currentMessage.length > 0)) {
+            super.publish(currentMessage);
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/app/templates/publisher.jsx]: shall we add a comment here?

File: product/distribution/carbon-home/generator-widgetgenerator/app/templates/publisher.jsx [hunk 25]
```diff
@@ -0,0 +1,333 @@
...
+        const { muiTheme } = this.props;
+        const themeName = muiTheme.name;
+        return (
+            <IntlProvider
+                locale={language}
+                messages={localeMessages}
+            >
+                <MuiThemeProvider
+                    theme={themeName === 'dark' ? darkTheme : lightTheme}
+                >
```
Comment (TanyaM) [inline product/distribution/carbon-home/generator-widgetgenerator/app/templates/publisher.jsx]: Shall we add a comment here like `sample code starts here` and another comment at where it ends saying `sample code ends here`

File: product/distribution/src/assembly/bin.xml [hunk 26]
```diff
@@ -90,6 +90,11 @@
             <directory>carbon-home/resources/security</directory>
             <outputDirectory>resources/security</outputDirectory>
             <fileMode>644</fileMode>
+        </fileSet>
+          <fileSet>
+            <directory>carbon-home/generator-widgetgenerator</directory>
```
Comment (TanyaM) [inline product/distribution/src/assembly/bin.xml]: As discussed offline let's make the directory name generator-widget

Comment (pahans) [reply product/distribution/src/assembly/bin.xml]: @TanyaM shall we make it `widget-generator`.

**Pull request merged**